vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt

```diff
@@ -0,0 +1,358 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.textnow.android.vessel
+
+import kotlinx.coroutines.CoroutineDispatcher
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.coroutineContext
+import kotlin.system.measureTimeMillis
+
+/**
+ * A time span over which some database operation occurred.
+ * Used in profiling database operations when using a [Vessel] instance.
+ * This shows where time is being spent when using [Vessel]
+ */
+enum class Span {
+    READ_FROM_DB,
+    WRITE_TO_DB,
+    PRELOAD_FROM_DB,
+    DELETE_FROM_DB,
+    REPLACE_IN_DB,
+    CLEAR_DB;
+
+    /**
+     * Friendly name for each enum - ex, "read", "write", etc.
+     */
+    fun niceName(): String {
+        return this.name.substring(0, this.name.indexOf("_")).lowercase()
+    }
+}
+
+/**
+ * A significant event that occurred when using a [Vessel] instance.
+ * Used to profile important events such as cache hits.
+ */
+enum class Event {
+    CACHE_HIT_READ,
+    CACHE_HIT_WRITE,
+    CACHE_HIT_DELETE,
+    CACHE_HIT_REPLACE;
+
+    /**
+     * Friendly name for each enum - ex, "read", "delete", etc.
+     */
+    fun niceName(): String {
+        return this.name.substring(this.name.lastIndexOf("_") + 1).lowercase()
+    }
+}
+
+/**
+ * A [Worker] is one of either a thread or a coroutine, and indicates where work is being done
+ */
+enum class WorkerType {
+    COROUTINE,
+    THREAD,
+}
+
+/**
+ * Tracks a worker - its [WorkerType] and name (thread or coroutine name)
+ */
+data class Worker(val type: WorkerType, val name: String)
+
+/**
+ * Tracks data for one [Span] - how many times it was entered, and how much time (total) has been
+ * spent inside this span
+ */
+data class SpanData(val span: Span, var hitCount: Int = 0, var totalDurationMS: Long = 0L)
+
+/**
+ * Tracks data for one [Event] - how many times the event has been hit
+ */
+data class EventData(val event: Event, var hitCount: Int = 0)
+
+/**
+ * Consolidate all profiling data for one [Worker] - the [Span]s and [Event]s executed by this
+ * [Worker].  In short, where this [Worker] is spending its time when using [Vessel].
+ */
+data class WorkerData(
+    val worker: Worker,
+    val spans: MutableMap<Span, SpanData> = mutableMapOf(),
+    val events: MutableMap<Event, EventData> = mutableMapOf(),
+) {
+    /**
+     * Total time spent by this [Worker], across all [Span]s
+     * That is, the time spent across all operations performed by this [Worker] in a [Vessel] instance
+     */
+    val totalDurationMS get() = spans.values.sumOf { it.totalDurationMS }
+}
+
+/**
+ * Helper to print text tables
+ */
+private fun table(
+    headings: List<Any>,
```

Comment:
do these need to be `Any` since you know the format?

Reply:
`headings` can be `String`.  The others can be anything that would be shown in table form - Ints, Enums, etc (usage here may just be Int, but I don't see a need to constrain the function further even if it is just used here).  

I updated `headings` but left the rest.